Issue/3589 menu selection bar db loader

mzorz opened the pull request:
Fixes #3589

This brings the Menu locations and Menus spinners to life. Now you can add/remove/edit menus and spinners will reflect so accordingly.

![add_edit_remove_menus3](https://cloud.githubusercontent.com/assets/6597771/15688981/3437576c-2753-11e6-8d04-40c4f58cb966.gif)

The way it works is, basically it loads the menus right away from the DB and renders the screen, and at the same time triggers a refresh call to the server. Once it comes back from the server, the DB is updated and the screen is updated as well.

Some things that need further discussion:
- check the default menu behaviour, is it coming in the API? As per my checks it is not, so I wonder whether there’s some magic ID number for the default menu if I wanted to edit the name or, in the very near future, the menu items. Is it possible to add/edit menu items on the Default Menu? How’s that handled?

- you’ll need to delete the database / clear app data / drop the Menus table programmatically once as I’ve added one missing column to it (SITE_ID_COLUMN)

Random notes:
+ added navigation bar to the activity
+ save info to db
+ tried to change look and feel of spinners to match the white background/blue border in the web
(tried and didn’t like it really, we’d better have @mattmiklic  take a look)

Further work:
- componentize the spinners as in the way MenuAddEditRemoveView works



Needs review: @tonyr59h 

**Added note**: please merge https://github.com/wordpress-mobile/WordPress-Android/pull/4146 **AFTER** reviewing and merging this one

tonyr59h commented:
Good job merging those branches @mzorz! The screen is shaping up nicely :)

![spinner](https://cldup.com/OgQmNiXlHi.png)

The spinners are almost finished, there are a few things I noticed:

* The TextView for the title needs to wrap_content, ideally shrinking the text size if needed so the spinner container stays constant
* The icons need to be a bit bigger to match the iOS and web counterparts
* This one is more of a question for @mattmiklic; the web looks like it uses semi-bold font for the title, but it looks normal on iOS (and in this PR). Should we stick with normal for mobile?

Some other design thoughts:

* Should `uses` be all caps or all lower case? Web uses all lower case, iOS is using all caps
* I think we should consider adding the drop-down arrow on the right hand side of the spinner
* Font size for the smaller spinner text (top text) needs to be increased and the top/bottom spinner padding looks excessive, Matt will probably weigh in on this with actual numbers
* The [Material guidelines](https://www.google.com/design/spec/components/buttons.html#buttons-button-types) hint that flat buttons should be favored over raised buttons. I think the Save and Trash buttons could be made flat
* I like that the edit icon is right next to the Menu name in the EditText [on iOS](https://hogwartsp2.files.wordpress.com/2016/04/menus-main6.png), should we consider re-sizing the EditText to wrap content?

Good call on hiding the Trash button for the default menu 👍

tonyr59h commented:
Ran into this crash when switching Menu Locations:

```
java.lang.IndexOutOfBoundsException: Invalid index 1, size is 1
    at java.util.ArrayList.throwIndexOutOfBoundsException(ArrayList.java:255)
    at java.util.ArrayList.get(ArrayList.java:308)
    at org.wordpress.android.ui.menus.MenusSpinner$MenusSpinnerAdapter.getView(MenusSpinner.java:134)
    at android.widget.Spinner.makeView(Spinner.java:617)
    at android.widget.Spinner.layout(Spinner.java:565)
    at android.widget.Spinner.onLayout(Spinner.java:527)
    at android.view.View.layout(View.java:15671)
    at android.view.ViewGroup.layout(ViewGroup.java:5038)
    at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1703)
    at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1557)
    at android.widget.LinearLayout.onLayout(LinearLayout.java:1466)
    at android.view.View.layout(View.java:15671)
    at android.view.ViewGroup.layout(ViewGroup.java:5038)
    at android.widget.LinearLayout.setChildFrame(LinearLayout.java:1703)
    at android.widget.LinearLayout.layoutVertical(LinearLayout.java:1557)
    at android.widget.LinearLayout.onLayout(LinearLayout.java:1466)
    at android.view.View.layout(View.java:15671)
    at android.view.ViewGroup.layout(ViewGroup.java:5038)
    at android.widget.FrameLayout.layoutChildren(FrameLayout.java:579)
    at android.widget.FrameLayout.onLayout(FrameLayout.java:514)
```

tonyr59h commented:
I'm able to lose the bottom spinner when switching between Menu Locations. Looks like this:

![](https://cldup.com/Uzy3bFF1j2.png)

mzorz commented:
Awesome feedback @tonyr59h ! Will tackle that crash and the UI glitches now

mzorz commented:
> Ran into this crash when switching Menu Locations:

Ok I added code to prevent the crash, but still wondering when such a situation is reached. That situation seems to happen when the adapter is re-set with each change to the Location spinner, but will have to further check. Let me know how it goes with the changes now.

> The TextView for the title needs to wrap_content, ideally shrinking the text size if needed so the spinner container stays constant

Fixed this and made the TextView a one-liner, with `ellipsize=end`.

> The icons need to be a bit bigger to match the iOS and web counterparts
Fixed

> This one is more of a question for @mattmiklic; the web looks like it uses semi-bold font for the title, but it looks normal on iOS (and in this PR). Should we stick with normal for mobile?

Letting @mattmiklic weigh in on this one

> Some other design thoughts:
> 
> Should uses be all caps or all lower case? Web uses all lower case, iOS is using all caps
> 

Setting to all-caps following iOS design, yet open to suggestions

> I think we should consider adding the drop-down arrow on the right hand side of the spinner

Will add it in a bit 

> Font size for the smaller spinner text (top text) needs to be increased and the top/bottom spinner padding looks excessive, Matt will probably weigh in on this with actual numbers

I just increased the smaller spinner text (top text) from 14sp to 16sp. Also decreased the padding from 16dp to 8dp.

> The Material guidelines hint that flat buttons should be favored over raised buttons. I think the Save and Trash buttons could be made flat

I don't think it's really enforced in terms of the material design guidelines, but we can surely do that if needed! My implementation follows the look the web shows, but I've seen it differently on iOS. Going to wait on @mattmiklic on this one

> I like that the edit icon is right next to the Menu name in the EditText on iOS, should we consider re-sizing the EditText to wrap content?

Can you provide a screenshot for this particular one please?

tonyr59h commented:
> Can you provide a screenshot for this particular one please?

Sure thing! Here's the iOS UI:
![](https://cldup.com/NsCCSZxpb4.png)

I'm referring to the pencil icon next to the menu name, it's next to the last character of the menu name on the web and iOS but at the end of the EditText on Android.

tonyr59h commented:
Note: Created #4156 to make sure we get design issues resolved.

mzorz commented:
> I'm referring to the pencil icon next to the menu name, it's next to the last character of the menu name on the web and iOS but at the end of the EditText on Android.

Fixed that one

> I'm able to lose the bottom spinner when switching between Menu Locations. 

fixed the second spinner disappearing bug :)

> I think we should consider adding the drop-down arrow on the right hand side of the spinner

added chevron / down arrow to the right of spinners too

mzorz commented on WordPress/src/main/java/org/wordpress/android/ui/menus/MenusFragment.java:
```diff
@@ -42,53 +56,208 @@ public void onCreate(Bundle bundle) {
+
+                        }.execute();
+
+                        addSpecialMenus(locations.get(0), menus);
```
I loved the handling of specialMenus here @tonyr59h ! much clearer 👍

mzorz commented:
Great work on encapsulating the special menu handling @tonyr59h !

I’ve been checking how the Android client works in this state against what the web does and noticed that as soon as I land on the Menus section on each one, I see this:
- both the Android app and web client show Top Header Menu selected in the Location spinner
- both the Android app and web show the Default menu selected in the Menus spinner

But, when having Top Header selected on the Locations spinner and changing it to the 3rd menu location (Social):
- the Web shows  “No Menu” in the Menus spinner
- the Android app shows  ‘Default Menu’ in the Menus spinner

A possible cause for this is that both Default Menu and No Menu occupy the zero position in the adapter array, so setSelection(0) after changing the Spinner's internal menu objects causes actually nothing to happen (because the adapter sees no change in position).
This happened to me before, and a workaround for this is to re-set the Spinner's adapter array (as per the lack of a more fine grained `notifyDataSetChanged()` method in SpinnerAdapter), which I’m doing now here:

https://github.com/wordpress-mobile/WordPress-Android/pull/4145/commits/ed4d785e1fc4ad32399eb26a8f37950c0cb4b83b

tonyr59h commented on WordPress/src/main/java/org/wordpress/android/ui/menus/MenusFragment.java:
```diff
@@ -42,53 +56,208 @@ public void onCreate(Bundle bundle) {
+                            item.details = menu.details;
+                            item.locations = menu.locations;
+                            item.menuItems = menu.menuItems;
+                            break;
```
^ That was all you, @mzorz 😄

mzorz replied:
really? LMFAO 😄

tonyr59h commented:
> This happened to me before, and a workaround for this is to re-set the Spinner's adapter

Works well, good catch 👍

tonyr59h commented:
Nice job taking care of those edge cases for Default Menu and No Menu too, I didn't notice that the view wasn't updated when switching from a Location with Default Menu to one with No Menu.

mzorz commented on WordPress/src/main/java/org/wordpress/android/models/MenuModel.java:
```diff
@@ -20,6 +20,10 @@
     public List<MenuLocationModel> locations;
     public List<MenuItemModel> menuItems;
 
+    public static final long ADD_MENU_ID = -1;
```
moved these constants from MenuLocationModel to MenuModel as they seemed to belong here

tonyr59h commented:
Selecting `Add Menu` works to create a new menu, but the spinner needs to be updated and the newly created Menu should be updated with the current location.

mzorz commented on WordPress/src/main/java/org/wordpress/android/ui/menus/MenusSpinner.java:
```diff
@@ -128,7 +128,7 @@ public View getView(int position, View convertView, ViewGroup parent) {
             }
 
             TextView titleView = (TextView) convertView.findViewById(R.id.menu_spinner_title);
-            titleView.setText(String.format(mTitle, getCount()));
```
oh - there's a problem with this, as the component is used for Menu Locations as well - maybe we can pass a `int baseCount` parameter in the creator method for MenuSpinner, or maybe in `setItems` that can be used as a subtractor?

mzorz replied:
Just adding a commit with this

tonyr59h replied:
Oops, missed that one! Good catch 👍

tonyr59h commented:
:shipit:
